src/blockchain/alt_block_tree.cpp

```diff
@@ -78,33 +54,172 @@ bool AltTree::bootstrapWithGenesis(ValidationState& state) {
   if (prev == nullptr) {
     return state.Invalid(
         "acceptBlockHeader()", "bad-prev-block", "can not find previous block");
   }
 
-  auto index = insertBlockHeader(block);
+  auto* index = insertBlockHeader(block);
 
   assert(index != nullptr &&
          "insertBlockHeader should have never returned nullptr");
 
-  addToChains(index);
+  // our current popState is not at desired "index->pprev" state, so change it
+  if (popState_ != index->pprev) {
+    AltBlockIndex* newPopState = popState_;
+    // we need to unapply payloads from this chain until the fork point,
+    // and then apply payloads from the fork chain
+    if (!unapplyAndApply(
+            pop_, &newPopState, *index, state, [this](AltBlockIndex* i) {
+              return this->prepo_->get(i->header.getHash());
+            })) {
+      pop_.rollback();
+      return state.addStackFunction("AltTree::acceptBlock");
+    }
+    popState_ = newPopState;
+  }
+
+  // we just add payloads via addPayloads, no need to unapply state
+  if (!pop_.addPayloads(payloads, state, stateChange)) {
+    return state.addStackFunction("AltTree::acceptBlock");
+  }
+  // current state has been changed
+  popState_ = index;
 
   return true;
 }
 
 bool AltTree::acceptBlock(const AltBlock& block,
                           const std::vector<Payloads>& payloads,
-                          ValidationState& state) {
+                          ValidationState& state,
+                          StateChange* change) {
   for (const auto& p : payloads) {
-    if (!acceptBlock(block, p, state)) {
+    if (!acceptBlock(block, p, state, change)) {
       return false;
     }
   }
 
   return true;
 }
+
+int AltTree::compareThisToOtherChain(AltBlockIndex* other) {
+  ValidationState state;
+
+  // create single-use pop manager, which internally combines payloads from
+  // both this chain and the other chain
+  PopManager combinedPop = pop_;
+  AltBlockIndex* combinedPopState = popState_;
+
+  auto height = config_->getBootstrapBlock().height;
+  bool ret = apply(
+      combinedPop, &combinedPopState, *other, state, [this](AltBlockIndex* i) {
+        return prepo_->get(i->header.getHash());
+      });
+
+  if (!ret) {
+    // other chain has bad payloads (expired?)
+    return 1;  // this chain wins
+  }
+
+  Chain<AltBlockIndex> thisChain(height, popState_);
+  Chain<AltBlockIndex> forkChain(height, other);
+
+  return combinedPop.compareTwoBranches(thisChain, forkChain);
+}
+
+void AltTree::unapply(
+    PopManager& pop,
+    AltBlockIndex** popState,
+    AltBlockIndex& to,
+    const std::function<std::vector<Payloads>(AltBlockIndex*)>& getPayloads) {
+  auto bootstrapHeight = config_->getBootstrapBlock().height;
+  Chain<AltBlockIndex> chain(bootstrapHeight, *popState);
+  auto* forkPoint = chain.findFork(&to);
+  auto* current = chain.tip();
+  while (current && current != forkPoint) {
+    // unapply payloads
+    for (const auto& p : getPayloads(current)) {
+      pop.removePayloads(p);
```

Comment:
Might be a good idea to have removePayloadsWithBlock function and avoid passing getPayloads functor here.